[KEP:4020] Peer-aggregated discovery

richabanker opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->
/kind feature

#### What this PR does / why we need it:
Modifies agg-discovery response to include  resources served by all apiservers in an HA cluster if UnknownVersionInteroperabilityProxy feature gate is enabled. Also accepts a `profile=nopeer` Accept header to request the non peer-aggregated (local only) view of discovery.

#### Which issue(s) this PR is related to:
<!--
Please link relevant issues to help with tracking.

To automatically close the linked issue(s) when this PR is merged,
add the word "Fixes" before the issue number or link.
Do not use "Fixes" if the PR is of kind `failing-test` or `flake`.

Reference KEPs when applicable in addition to specific issues.

Examples:
Fixes #<issue number>
<issue link> (issue in a different repository)
KEP: https://github.com/kubernetes/enhancements/issues/<kep-issue-number>

If there is no associated issue, then write "N/A".
-->
https://github.com/kubernetes/enhancements/issues/4020

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Enhanced discovery response to support merged API groups/resources from all peer apiservers when UnknownVersionInteroperabilityProxy feature is enabled
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

hashim21223445 reviewed (commented):
[](url)

jpbetz commented on pkg/controlplane/apiserver/server.go:
```diff
@@ -230,6 +230,18 @@ func (c completedConfig) New(name string, delegationTarget genericapiserver.Dele
+					return fmt.Errorf("MergedDiscoveryHandler not configured")
+				}
+
+				s.GenericAPIServer.Handler.NonGoRestfulMux.HandleFunc("/api/merged", mergedHandler.ServeHTTP)
+				s.GenericAPIServer.Handler.NonGoRestfulMux.HandleFunc("/apis/merged", mergedHandler.ServeHTTP)
```
Merged discovery isn't a new sub-resource, but rather a different representation of the existing /api and /apis resources.

My understanding is that the returned Kind would be the same in both cases. The key difference is that the existing resources return a single level of data, while a merged view would return all levels.

We have a precedent for requesting different representations with `Accept: application/json;as=Table;v=v1beta1;g=meta.k8s.io`. However, we can't use the as parameter here, since discovery would use `as=APIResourceDiscovery;v=v2beta1;g=apidiscovery.k8s.io` for both the normal and merged views.

What about using a different parameter, like `application/json;profile=Merged` or `application/json;scope=All`?

jpbetz replied:
Per offline discussion, I think we can returned merged discovery by default when the feature is on, and only serve unmerged discovery when peers specifically ask for it.  This way, discovery and the APIs actually made available by an apiserver to agree.

The fallback here is important.  For clients that do provide `profile=Unmerged` (or whatever we decide to call it), if the server doesn't support merged discovery, it returned unmerged data, and if the server *does* support merged discovery, it will respect the param and served unmerged data.  So a Accept header works great.

jpbetz commented on pkg/apis/apidiscovery/types.go:
```diff
@@ -110,6 +110,12 @@ type APIResourceDiscovery struct {
+	// are capable of handling requests for this resource type.
+	// +listType=set
+	// +optional
+	ServerIDs []string `json:"serverIds,omitempty"`
```
How does this get populated?  Is there a chicken-egg problem in the information served here?  Would it be possible to implement UVIP/MVP without this field?

richabanker commented:
/ok-to-test

liggitt commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/negotiation/negotiate.go:
```diff
@@ -227,6 +230,10 @@ func acceptMediaTypeOptions(params map[string]string, accepts *runtime.Serialize
 
+		// controls the discovery profile (merged vs unmerged)
+		case "profile":
+			options.Profile = v
```
> that is what keeps merged-discovery working for older apiservers (unaware of the feature) doesn't it? They silently accept the `unmerged` header and return the original (un-merged) discovery? This silent acceptance is what enables backward compatibility: clients like MVP can request `profile=unmerged`, and older servers will still respond as usual, without error.

What I *expected* to keep things compatible was:

1. Old clients would keep requesting `Accept: application/json;g=apidiscovery.k8s.io;v=v2;as=APIGroupDiscoveryList, application/json`
2. New clients wanting "unmerged falling back to normal" would include multiple things in their Accept header: `Accept: application/json;g=apidiscovery.k8s.io;v=v2;as=APIGroupDiscoveryList;profile=unmerged, application/json;g=apidiscovery.k8s.io;v=v2;as=APIGroupDiscoveryList, application/json`
3. New clients wanting "unmerged erroring if unsupported" would only include the request for unmerged in their Accept header: `Accept: application/json;g=apidiscovery.k8s.io;v=v2;as=APIGroupDiscoveryList;profile=unmerged`

For that to work, old servers would have to skip the unrecognized "profile=unmerged" bit in option 2 to successfully serve one of the later content-types, but would have to error on that in option 3 if that was the only content type the client indicated they accept

Maybe we don't have any clients like 3, so maybe it doesn't matter.

richabanker replied:
Ack'd, this helps clarify my confusion, thanks.
I could attempt to have a way to error for Option 3, I'll get back once I have had some time to think about this more.

richabanker commented on staging/src/k8s.io/apiserver/pkg/endpoints/discovery/aggregated/handler.go:
```diff
@@ -517,58 +582,169 @@ func (rdm *resourceDiscoveryManager) ServeHTTP(resp http.ResponseWriter, req *ht
+
+	// Process local groups first.
+	for _, group := range localGroups {
+		groupCopy := group.DeepCopy()
```
Actually, at the time of processing localGroups, we are not really modifying the group, so not sure if this DeepCopy() can be avoided at this point.

richabanker commented on staging/src/k8s.io/apiserver/pkg/endpoints/discovery/aggregated/handler.go:
```diff
@@ -517,58 +582,169 @@ func (rdm *resourceDiscoveryManager) ServeHTTP(resp http.ResponseWriter, req *ht
+		result = append(result, *group)
+	}
+
+	// TODO: is this right?
```
Ack'd finally was able to make CI happy, will get to this next. Keeping open.

richabanker replied:
That looks amazing, incorporated! Thank you so much for drafting out a neat solution preserving the relative ordering between groups like that.

richabanker commented on staging/src/k8s.io/client-go/discovery/discovery_client.go:
```diff
@@ -59,12 +59,15 @@ const (
-	AcceptV2Beta1 = runtime.ContentTypeJSON + ";" + "g=apidiscovery.k8s.io;v=v2beta1;as=APIGroupDiscoveryList"
-	AcceptV2      = runtime.ContentTypeJSON + ";" + "g=apidiscovery.k8s.io;v=v2;as=APIGroupDiscoveryList"
+	AcceptV2Beta1         = runtime.ContentTypeJSON + ";" + "g=apidiscovery.k8s.io;v=v2beta1;as=APIGroupDiscoveryList"
+	AcceptV2Beta1Unmerged = runtime.ContentTypeJSON + ";" + "g=apidiscovery.k8s.io;v=v2beta1;as=APIGroupDiscoveryList;profile=unmerged"
```
Done.

Jefftree commented on staging/src/k8s.io/apimachinery/pkg/util/sort/sort.go:
```diff
@@ -0,0 +1,91 @@
+// we use this function to compute a consensus ordering that respects the relative orderings
+// reported by all peers. This is called from the merged discovery handler when constructing
+// the final APIGroupDiscoveryList for clients.
+func SortDiscoveryGroupsTopo(peers [][]string) []string {
```
wow I love this, solves our ordering problem really nicely!

richabanker replied:
This is @liggitt 's idea, and yeah fits quite nicely!

richabanker commented on staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:
```diff
@@ -859,7 +865,8 @@ func (s *GenericAPIServer) InstallLegacyAPIGroup(apiPrefix string, apiGroupInfo
 	legacyRootAPIHandler := discovery.NewLegacyRootAPIHandler(s.discoveryAddresses, s.Serializer, apiPrefix)
-	wrapped := discoveryendpoint.WrapAggregatedDiscoveryToHandler(legacyRootAPIHandler, s.AggregatedLegacyDiscoveryGroupManager)
+	// No peer-to-peer discovery for generic apiserver.
+	wrapped := discoveryendpoint.WrapAggregatedDiscoveryToHandler(legacyRootAPIHandler, s.AggregatedLegacyDiscoveryGroupManager, s.AggregatedLegacyDiscoveryGroupManager)
```
We aren't, I passed the `AggregatedLegacyDiscoveryGroupManager` handler for the arg that expects peerMergedHandler

liggitt commented on staging/src/k8s.io/apiserver/pkg/server/config.go:
```diff
@@ -1123,7 +1123,8 @@ func installAPI(name string, s *GenericAPIServer, c *Config) {
 
 	if c.EnableDiscovery {
-		wrapped := discoveryendpoint.WrapAggregatedDiscoveryToHandler(s.DiscoveryGroupManager, s.AggregatedDiscoveryGroupManager)
+		// No peer-to-peer discovery for generic apiservers.
```
pass in s.PeerAggregatedDiscoveryManager so generic servers *can* provide a handler in their config if they want

in WrappedHandler#ServeHTTP, use aggHandler if peerAggHandler is not provided

```diff
@@ -49,9 +50,11 @@ func (wrapped *WrappedHandler) ServeHTTP(resp http.ResponseWriter, req *http.Req
 				wrapped.aggHandler.ServeHTTP(resp, req)
 				return
 			}
-			// Serve peer-aggregated discovery by default.
-			wrapped.peerAggHandler.ServeHTTP(resp, req)
-			return
+			// Serve peer-aggregated discovery by default if a peer handler is available.
+			if wrapped.peerAggHandler != nil {
+				wrapped.peerAggHandler.ServeHTTP(resp, req)
+				return
+			}
 		}
 		wrapped.aggHandler.ServeHTTP(resp, req)
 		return

```

liggitt commented on staging/src/k8s.io/apimachinery/pkg/util/sort/sort.go:
```diff
@@ -0,0 +1,91 @@
+// we use this function to compute a consensus ordering that respects the relative orderings
+// reported by all peers. This is called from the peer-aggregated discovery handler when constructing
+// the final APIGroupDiscoveryList for clients.
+func SortDiscoveryGroupsTopo(peers [][]string) []string {
```
I think the order of peers can affect the results, so should document that callers should pass in peers in a consistent order to get deterministic results, and when calling this, we should sort peers by something that all peers will have access to (identity name or something)

liggitt commented on staging/src/k8s.io/apimachinery/pkg/util/sort/sort.go:
```diff
@@ -0,0 +1,91 @@
+// we use this function to compute a consensus ordering that respects the relative orderings
+// reported by all peers. This is called from the peer-aggregated discovery handler when constructing
+// the final APIGroupDiscoveryList for clients.
+func SortDiscoveryGroupsTopo(peers [][]string) []string {
```
nit: there's nothing really discovery-specific in the algorithm, maybe rename this to something more generic like `MergePreservingSharedRelativeOrder`, etc

richabanker replied:
Renamed to `MergePreservingRelativeOrder`

liggitt commented on staging/src/k8s.io/apiserver/pkg/endpoints/discovery/aggregated/wrapper.go:
```diff
@@ -28,18 +28,31 @@ import (
+			}
+			// Serve peer-aggregated discovery by default.
+			wrapped.peerAggHandler.ServeHTTP(resp, req)
+			return
```
allow peerAggHandler to be nil and fallback to aggHandler, so generic API servers can propagate their peer handler if they want and this will do something sensible if not set (and so kube-apiserver can just leave peerAggHandler nil if it doesn't have a peer proxy)

```suggestion
			if wrapped.peerAggHandler != nil {
				// Serve peer-aggregated discovery by default if provided.
				wrapped.peerAggHandler.ServeHTTP(resp, req)
				return
			}
```

liggitt commented on staging/src/k8s.io/apiserver/pkg/endpoints/discovery/aggregated/peer_aggregated_handler.go:
```diff
@@ -0,0 +1,254 @@
+
+func (h *peerAggDiscoveryHandler) mergeResources(
+	localGroups []apidiscoveryv2.APIGroupDiscovery,
+	peerGroupDiscovery map[string][]apidiscoveryv2.APIGroupDiscovery,
```
how are peers ordered? what is the key of this map? I think we need to deterministically order our arguments to `SortDiscoveryGroupsTopo` to get consistent results, so accumulating the groups / versions we pass in by something like identity (including ordering our local values by our own local identity) might be a possibility

richabanker replied:
key is the serverID. You're right, without passing in consistent ordered slices, we risk each server constructing a different ordering of the groups. I am now doing the following:
1. create a combined map from serverID -> []APIGroupDiscovery using both local and peer servers 
2. sort the keys of of this combined map and store them in a list
3. iterate over the sorted IDs list to process each server's []APIGroupDiscovery one by one

liggitt commented on staging/src/k8s.io/apiserver/pkg/util/peerproxy/peerproxy_handler.go:
```diff
@@ -152,10 +161,6 @@ func (h *peerProxyHandler) WrapHandler(handler http.Handler) http.Handler {
 
 		gvr := schema.GroupVersionResource{Group: requestInfo.APIGroup, Version: requestInfo.APIVersion, Resource: requestInfo.Resource}
-		if requestInfo.APIGroup == "" {
-			gvr.Group = "core"
```
was this bogus before?

liggitt commented on staging/src/k8s.io/apiserver/pkg/util/peerproxy/peer_discovery.go:
```diff
@@ -89,63 +96,85 @@ func (h *peerProxyHandler) syncPeerDiscoveryCache(ctx context.Context) error {
+			if groupDiscovery.Name == "" {
+				continue
+			}
+			groupMap[groupDiscovery.Name] = groupDiscovery
```
doesn't using a map here lose ordering we care about?

richabanker replied:
switched to appending to list directly instead of map.

richabanker commented on staging/src/k8s.io/apiserver/pkg/endpoints/discovery/aggregated/peer_aggregated_handler.go:
```diff
@@ -0,0 +1,254 @@
+	if len(peerGroupDiscovery) == 0 {
+		return localGroups
+	}
+
```
Hmm yeah I was doing the merging for groups, versions, resources unconditionally... tried to refactor it now. The changes are as follows 

1. to have deterministic ordering of the final []APIGroupDiscovery, I combined the local and peer info into a single map keyed by serverID and sorted the keys. Then iterate over the keys to perform/check for merging.

2. For each iteration, mergeVersionsAcrossGroups() is called (which also calls mergeResourcesAcrossVersions()) but enhanced them to not unconditionally merge 
     1. They return a boolean flag `didMerge` that is true only if new content was found or the order was different
     2. If no changes are detected (the lists are identical), the function bails out early
3. this `didMerge` bubbles up for each server, and the top level mergeResources() checks if all `didMerge` 's reported true and no change in groups order was seen (checked using reflect.DeepEqual) - it also skips merging

richabanker commented on staging/src/k8s.io/apiserver/pkg/util/peerproxy/peer_discovery.go:
```diff
@@ -89,63 +96,85 @@ func (h *peerProxyHandler) syncPeerDiscoveryCache(ctx context.Context) error {
 		}
 	}
 
+	groupList := make([]apidiscoveryv2.APIGroupDiscovery, 0, len(groupMap))
```
switched to appending to list directly instead of map.

richabanker commented on staging/src/k8s.io/apiserver/pkg/endpoints/discovery/aggregated/handler.go:
```diff
@@ -138,16 +145,30 @@ type resourceDiscoveryManager struct {
+	lock                  sync.RWMutex
+	apiGroups             map[groupKey]*apidiscoveryv2.APIGroupDiscovery
+	versionPriorities     map[groupVersionKey]priorityInfo
+	invalidationCallbacks []func()
```
Done

richabanker commented on staging/src/k8s.io/apiserver/pkg/endpoints/discovery/aggregated/wrapper.go:
```diff
@@ -68,10 +83,10 @@ func (wrapped *WrappedHandler) GenerateWebService(prefix string, returnType inte
 // emit the aggregated discovery by passing in the aggregated
 // discovery type in content negotiation headers: eg: (Accept:
 // application/json;v=v2;g=apidiscovery.k8s.io;as=APIGroupDiscoveryList)
```
Done

richabanker commented:
@liggitt Replying to [your prev comment](https://github.com/kubernetes/kubernetes/pull/134514#discussion_r2470240243)

> I thought we would be updating the cache per-peer, with per-peer additions/updates/removals triggered by that peer's lease or endpoints getting added / removed / updated.

we [decided](https://github.com/kubernetes/kubernetes/pull/128402#discussion_r1993352137) in the implementation of the peer_discovery controller to have a level based logic - so in the sync loop, always get the list of coordination.v1 identity leases that exist in that moment in time and work on driving those to the desired state.

For Part 1 - Are we triggering updates on all of these things:
1. watched add/update/delete events of the identity leases: 
yes [here](https://github.com/kubernetes/kubernetes/blob/06bbf2abbcdaf3e1a4f6974e8bc32baba5b9587a/staging/src/k8s.io/apiserver/pkg/util/peerproxy/peerproxy.go#L115-L141)

2. polled add/update/delete of the peer endpoints: 
indirectly yes - when the peer identity lease informer does a periodic refresh(every 10m) and finds peer leases for each of which it tries to fetch discovery using the latest endpoints for each peer - if it fails to fetch discovery for any reason(like endpoint being deleted), we drop that peer from peer-agg discovery

3. something that makes us fetch peer discovery anew when CRD-defined types, or APIService-defined group/versions get removed (otherwise I think we'll keep listing them until our next periodic refresh of peers): 
we only refresh the local apiserver's discovery periodically([30m](https://github.com/kubernetes/kubernetes/blob/06bbf2abbcdaf3e1a4f6974e8bc32baba5b9587a/staging/src/k8s.io/apiserver/pkg/util/peerproxy/peerproxy.go#L48)) [here](https://github.com/kubernetes/kubernetes/blob/06bbf2abbcdaf3e1a4f6974e8bc32baba5b9587a/staging/src/k8s.io/apiserver/pkg/util/peerproxy/peerproxy_handler.go#L198-L202) and for peer discovery rely on the periodic refresh of the lease informer which is created [here](https://github.com/kubernetes/kubernetes/blob/06bbf2abbcdaf3e1a4f6974e8bc32baba5b9587a/pkg/controlplane/apiserver/config.go#L325). The default resync time for this informer being [10m](https://github.com/kubernetes/kubernetes/blob/06bbf2abbcdaf3e1a4f6974e8bc32baba5b9587a/pkg/controlplane/apiserver/config.go#L160)
4. periodic refreshes on some interval:
same as above

For Part 2- Are we dropping previously fetched peer discovery data for a particular peer out of the cache for all of these reasons:
1. no endpoint exists for the peer: yes - either when its identity lease also gets deleted or during the periodic resync of the identity lease informer we try to fetch discovery for it and fail
2. no identity lease exists for the peer: yes, we remove the corresponding discovery info when the delete-lease event is observed or during the resync of the identity lease informer
3. eventual repeated failure to fetch from a peer that does have endpoints+lease (some number of attempts with backoff or something): yes [here](https://github.com/kubernetes/kubernetes/blob/06bbf2abbcdaf3e1a4f6974e8bc32baba5b9587a/staging/src/k8s.io/apiserver/pkg/util/peerproxy/peer_discovery.go#L183-L198) but this is also done when we do a resync of the peer identity leases or observe any change in them

richabanker commented on staging/src/k8s.io/apiserver/pkg/util/peerproxy/exclusion_filter.go:
```diff
@@ -0,0 +1,329 @@
+	return filteredDiscovery
+}
+
+// markGVForDeletion checks if a group-version is still in use.
```
done

jpbetz commented on staging/src/k8s.io/apimachinery/pkg/util/sort/sort.go:
```diff
@@ -0,0 +1,102 @@
+
+func pathFrom(from, to string, links map[string]sets.Set[string]) bool {
+	visited := sets.New[string]()
+	tovisit := sets.New(from)
```
```suggestion
	toVisit := sets.New(from)
```

jpbetz commented on staging/src/k8s.io/apimachinery/pkg/util/sort/sort.go:
```diff
@@ -0,0 +1,102 @@
+				return true
+			}
+			if !visited.Has(next) {
+				tovisit.Insert(next)
```
nit: Don't need to guard set insertion with a `Has` check.

jpbetz commented on staging/src/k8s.io/apimachinery/pkg/util/sort/sort.go:
```diff
@@ -0,0 +1,102 @@
+func pathFrom(from, to string, links map[string]sets.Set[string]) bool {
+	visited := sets.New[string]()
+	tovisit := sets.New(from)
+	for {
```
What's the time complexity of this?  Can we verify it with a bench?  I was expecting an algorithm where two lists are merged via an O(n) parallel traverse, and N lists are merged by merging repeatedly.  I'm having trouble convincing myself that this is more efficient.

jpbetz reviewed (commented):
I got hung up on the sort algorithm. Will try to get back to this later.

jpbetz commented on staging/src/k8s.io/apimachinery/pkg/util/sort/sort.go:
```diff
@@ -0,0 +1,102 @@
+//   - Input {{"b", "c", "a"}, {"a", "b", "c"}} might return {"b", "c", "a"}.
+//
+// This is useful for creating a stable, consistent ordering when merging data from
+// multiple sources that may have partial but not conflicting orderings.
```
Thanks, I find this much easier to reason about.

jpbetz commented on staging/src/k8s.io/apiserver/pkg/endpoints/discovery/aggregated/peer_aggregated_handler.go:
```diff
@@ -0,0 +1,483 @@
+
+// mergeVersionsAcrossGroup merges two APIGroupDiscovery objects.
+// It returns a new merged object and a boolean indicating if any
+// new versions/resources were added or if the order was different.
```
```suggestion
// It returns a new merged object and a boolean indicating if the result differs from a.
// The result may differ if any new versions/resources were added or if the order was different.
```

?

jpbetz commented on staging/src/k8s.io/apiserver/pkg/endpoints/discovery/aggregated/peer_aggregated_handler.go:
```diff
@@ -0,0 +1,483 @@
+
+// mergeResourcesAcrossVersion merges two APIVersionDiscovery objects.
+// It returns a new merged object and a boolean indicating if any
+// new resources were added or if the order was different.
```
Same here

jpbetz commented on staging/src/k8s.io/apiserver/pkg/endpoints/discovery/aggregated/peer_aggregated_handler_test.go:
```diff
@@ -0,0 +1,423 @@
+limitations under the License.
+*/
+
+package aggregated_test
```
Sorry to ask but could we add unit test function that specifically unit test `fetchFromCache`, `mergeResources` and the other non-trivial functions declared in `peer_aggregated_handler.go` in isolation?  I'd find that helpful in confirming my understanding of those functions.

richabanker replied:
tests for mergeResources are already there, added a new `TestFetchFromCache()` for `fetchFromCache `

jpbetz commented on staging/src/k8s.io/apiserver/pkg/util/peerproxy/exclusion_filter.go:
```diff
@@ -0,0 +1,370 @@
+	}
+
+	<-ctx.Done()
+	klog.Info("Shutting down GV deletion workers")
```
nit: I found it the way these lines combine with `defer h.gvDeletionQueue.ShutDown()` tricky to read.

Maybe above do:

```go
defer func() {
  klog.Info("Shutting down GV deletion workers")
  h.gvDeletionQueue.ShutDown()
}()
```

And then here do:

```
<-ctx.Done() // Wait till context is done to call deferred shutdown function
```

?

I'm assuming that we want to call shutdown immediately if there is a panic raised in this function..

jpbetz commented on staging/src/k8s.io/apiserver/pkg/util/peerproxy/exclusion_filter.go:
```diff
@@ -0,0 +1,370 @@
+	}
+
+	h.excludedGVsMu.Lock()
+	defer h.excludedGVsMu.Unlock()
```
Do we need to lock the call to `isGVUsed` ?  If so can we document that isDVUsed is unsafe?  If not, can we narrow down what code we hold the lock for?

richabanker replied:
We dont need to lock for `isGVUsed` since it only invoked a list() on informers. But since acquiring the lock (which happens after the first call to `isGVUsed`) may take some time, we invoke`isGVUsed` once again to be sure we are looking at the latest state, to avoid any race. 
If instead, we first acquire the lock and directly call `isGVUsed()` once after that - that may cause the lock to be help up for a long time since `isGVUsed()` does a list() on both crd and apiservice informer.

richabanker replied:
Maybe that race is super unlikely so it maybe ok to just not do the `isGVUsed(gv)` a second time? Happy to remove it if you think its an overkill

jpbetz commented on staging/src/k8s.io/apiserver/pkg/util/peerproxy/local_discovery.go:
```diff
@@ -0,0 +1,100 @@
+		for {
+			select {
+			case <-h.localDiscoveryCacheTicker.C:
+				klog.V(4).Infof("Invalidating local discovery cache")
```
Just double checking: Is the use of `V(4)` only on this log line (and not on any of the others) in this function intentional?

jpbetz commented on staging/src/k8s.io/apiserver/pkg/util/peerproxy/peer_discovery.go:
```diff
@@ -38,7 +38,14 @@ import (
+	// With the default rate limiting (exponential backoff starting at 5ms), 20 retries
+	// provides approximately 60-90 seconds of retry window, which is sufficient for
+	// both identity and endpoint leases to be established during normal startup.
+	maxRetries = 20
```
So this is 20 retries _per_ lease update? Meaning that so long as the peer is updating the lease regularly, and eventually is able to serve a successful response, that it will eventually be included in peer-discovery?

If so, can we clarify in the comment that this retry limit shouldn't result in indefinite peer exclusion?

jpbetz commented on staging/src/k8s.io/apiserver/pkg/util/peerproxy/peer_discovery.go:
```diff
@@ -196,3 +238,74 @@ func (h *peerProxyHandler) handleErr(err error, key string) {
+			return nil, false
+		}
+
+		if lease.Labels == nil || lease.Labels[identityLeaseLabel[0]] != identityLeaseLabel[1] {
```
`lease.Labels[identityLeaseLabel[0]]` returns `""` if `identityLeaseLabel[0]`  is not in the `lease.Labels` map. Is that safe?  Should we do an explicit check? Or are we confident that `""` will never be used as a valid value?

richabanker committed:
util/sort: Add MergePreservingRelativeOrder for topological sorting

Add a topological sort implementation that merges multiple ordered lists
while preserving their relative ordering. This is used by peer-aggregated
discovery to deterministically merge discovery documents from multiple
API servers.

Part of KEP-4020: Unknown Version Interoperability Proxy

richabanker committed:
discovery: Add profile-based content negotiation

Add support for 'profile=nopeer' in Accept headers to allow clients
to opt out of peer-aggregated discovery and request local-only results.

Updates discovery client to set appropriate Accept headers based on
whether peer-aggregated discovery is desired.

Part of KEP-4020: Unknown Version Interoperability Proxy

richabanker committed:
apiserver: Add peer proxy and peer-aggregated discovery support

Integrate peerproxy and peer-aggregated discovery into generic API server:
- Add PeerProxyConfiguration to server config
- Wire up peer discovery cache sync as post-start hook
- Connect local discovery invalidation callbacks
- Add GV deletion workers and reaper for safe cleanup

Part of KEP-4020: Unknown Version Interoperability Proxy

richabanker committed:
discovery/aggregated: Add peer-aggregated discovery handler

Implement peer-aggregated discovery that merges local and peer discovery
documents into a unified view. Features:
- Deterministic merging using topological sort
- Short-circuit optimizations when content/order unchanged
- Resource capability merging (verbs, subresources)
- ETag-based caching
- Prometheus metrics for cache hits/misses

The wrapper routes requests to either local-only or peer-aggregated
handlers based on Accept header profile.

Part of KEP-4020: Unknown Version Interoperability Proxy

richabanker committed:
discovery/aggregated: Add invalidation callback support

Add callback mechanism to notify peer-aggregated discovery handler
when local discovery changes, enabling efficient cache invalidation.

Part of KEP-4020: Unknown Version Interoperability Proxy

richabanker committed:
peerproxy: Refactor into separate files and add exclusion filter

richabanker committed:
kube-apiserver: Enable peer proxy and peer-aggregated discovery

Wire up peer proxy infrastructure in kube-apiserver:
- Add UnknownVersionInteroperabilityProxy feature gate
- Configure peer proxy with identity lease selectors
- Register CRD and APIService informers with exclusion filters
- Start peer discovery sync and GV cleanup workers

Includes extractors for CRDs and APIServices to identify which
GroupVersions should be excluded from peer discovery.

Part of KEP-4020: Unknown Version Interoperability Proxy

richabanker committed:
test: Add integration tests for peer-aggregated discovery

Add comprehensive integration tests covering:
- Peer-aggregated discovery merging across multiple API servers
- CRD/APIService exclusion from peer discovery
- Profile-based content negotiation (nopeer)
- Cache invalidation on local discovery changes
- Request routing to peer servers

Part of KEP-4020: Unknown Version Interoperability Proxy

richabanker commented on pkg/controlplane/apiserver/server.go:
```diff
@@ -213,22 +212,34 @@ func (c completedConfig) New(name string, delegationTarget genericapiserver.Dele
+				return nil
+			})
+
+			// Run excluded-groups reaper loop
```
Done

richabanker commented on staging/src/k8s.io/apiserver/pkg/endpoints/discovery/aggregated/peer_aggregated_handler.go:
```diff
@@ -0,0 +1,483 @@
+
+// PeerAggResourceManager defines the interface for managing peer-aggregated discovery resources
+// that combines both local and peer server resources.
+type PeerAggResourceManager interface {
```
Done

richabanker commented on staging/src/k8s.io/apiserver/pkg/util/peerproxy/peerproxy.go:
```diff
@@ -33,19 +33,29 @@ import (
+	// Local discovery cache needs to be refreshed periodically to store
+	// updates made to custom resources or aggregated resource that can
+	// change dynamically.
+	localDiscoveryRefreshInterval = 30 * time.Minute
```
I tried to enhance the comment. This is just the frequency at which we will update the local-discovery cache which is only used by mixed-version-proxy to decide whether a resource request should be handled locally or routed to a peer. It doesnt inform/influence the discovery serving path at all (clients see new CRDs/APIs immediately in discovery responses). 

Its more of a due-diligence measure for mvp to ensure that the local-discovery data its using to make routing decision is getting updated at a regular interval.

You might wonder what happens if we only refresh every 30 minutes—wouldn't that be problematic for newly created CRDs? That's handled separately: the exclusion_filter that we added in this PR ensures that CRDs and aggregated APIs are never included in the peer-discovery cache. This means the mvp has no choice but to let the local API server handle those requests, which is exactly what we want for CRDs and agg APIs. So the 30-minute refresh only impacts routing decisions for built-in APIs, not CRDs/aggregated APIs.

jpbetz replied:
Godoc update looks good, I was seeing stale data.

jpbetz commented on staging/src/k8s.io/apiserver/pkg/endpoints/discovery/aggregated/handler.go:
```diff
@@ -521,53 +540,62 @@ func (rdm *resourceDiscoveryManager) serveHTTP(resp http.ResponseWriter, req *ht
+	writeDiscoveryResponse(&response, etag, rdm.serializer, resp, req)
+}
+
+func writeDiscoveryResponse(
```
Is introducing this function needed in this PR?  I looks to me like it would fine to leave this code in `serveHTTP`, and would make it sightly easier to review this PR (any non-gate changes need to be reviewed VERY carefully, and this adds noise)

jpbetz replied:
Ah, I missed the new call site. nvm.

Jefftree commented on staging/src/k8s.io/apiserver/pkg/util/peerproxy/exclusion_filter.go:
```diff
@@ -0,0 +1,375 @@
+			if gvs == nil {
+				return
+			}
+			h.addExcludedGVs(gvs)
```
I originally didn't comment that we need a queue here because the operation looked really simple and fast. On a second look, `addExcludedGVs` requires holding a lock. it's possible for `markGVForDeletionIfUsed` to hold the lock and call `isGVUsed`. That operation calls a LIST on both CRDs and APIServices. `addExcludedGVs` would block until the list check is completed. That isn't ideal in eventhandlers, would it be doable to refactor this to use a queue?

jpbetz commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go:
```diff
@@ -275,7 +280,18 @@ func (c completedConfig) NewWithDelegate(delegationTarget genericapiserver.Deleg
+		s.GenericAPIServer.PeerAggregatedDiscoveryManager = aggregated.NewPeerAggregatedDiscoveryHandler(s.GenericAPIServer.APIServerID, s.GenericAPIServer.AggregatedDiscoveryGroupManager, c.ExtraConfig.PeerProxy, "apis")
+
+		// Register cache invalidation callbacks to the peer proxy handler
+		if c.ExtraConfig.PeerProxy != nil {
```
Is this check redundant given line 283?

Jefftree commented:
/lgtm

Thanks Richa!

liggitt commented:
/skip
/milestone v1.35
